### PR TITLE
kv/bulk: add guardrails to createSplitSSTable

### PR DESCRIPTION
**Pull request by msbutler**
Previously, CreateSplitSSTable could return an empty lhs or rhs sstable if the user provided a bad split point. This patch adds guardrails to prevent this and a unit test for this function.

Informs #117739

Release note: none

### CODE REVIEW

File: pkg/kv/bulk/split_sst_test.go [hunk 1]
```diff
@@ -0,0 +1,118 @@
...
+		},
+	} {
+		t.Run(tc.name, func(t *testing.T) {
+			iter, err := storage.NewMemSSTIterator(sstBytes, true, iterOpts)
+			require.NoError(t, err)
+			defer iter.Close()
+
+			iter.SeekGE(storage.MVCCKey{Key: start})
+			sendStart := timeutil.Now()
+			stats, err := storage.ComputeStatsForIter(iter, sendStart.UnixNano())
```
Comment (msbutler) [inline pkg/kv/bulk/split_sst_test.go:94]: one surprise here: `ComputStatsIter` eventually invalidates the iter, yet `createSplitSSTable` can reuse it by seeking to the start point again.

File: pkg/kv/bulk/split_sst_test.go [hunk 2]
```diff
@@ -0,0 +1,121 @@
...
+	"github.com/cockroachdb/cockroach/pkg/util/timeutil"
+	"github.com/stretchr/testify/require"
+)
+
+func TestSplitSST(t *testing.T) {
+	defer leaktest.AfterTest(t)()
+	defer log.Scope(t).Close(t)
+	ctx := context.Background()
+
+	s, _, _ := serverutils.StartServer(t, base.TestServerArgs{})
```
Comment (stevendanna) [inline pkg/kv/bulk/split_sst_test.go:34]: [nit] If you are just using this for cluster settings then you can do:

```
	st := cluster.MakeTestingClusterSettings()
```

instead

File: pkg/kv/bulk/sst_batcher.go [hunk 3]
```diff
@@ -1001,6 +988,30 @@ func createSplitSSTable(
...
 		return nil, nil, err
 	}
+	if !split {
+		return nil, nil, errors.Newf("split key %s after last key %s", last.Next())
+	}
 	right = &sstSpan{start: first, end: last.Next(), sstBytes: sstFile.Data()}
 	return left, right, nil
 }
+
+func addStatsToSplitTables(left, right, original *sstSpan, sendStartTimestamp time.Time) error {
```
Comment (stevendanna) [inline pkg/kv/bulk/sst_batcher.go:998]: I might put a small comment on the strategy here of computing the stats for one side and reusing the original stats to calculate the remainder.

File: pkg/kv/bulk/sst_batcher.go [hunk 4]
```diff
@@ -1001,6 +988,30 @@ func createSplitSSTable(
 	if err != nil {
 		return nil, nil, err
 	}
+	if !split {
+		return nil, nil, errors.Newf("split key %s after last key %s", last.Next())
```
Comment (stevendanna) [inline pkg/kv/bulk/sst_batcher.go:992]: Two format placeholders but only one argument.

File: pkg/kv/bulk/sst_batcher.go [hunk 5]
```diff
@@ -950,6 +934,9 @@ func createSplitSSTable(
 	settings *cluster.Settings,
 ) (*sstSpan, *sstSpan, error) {
 	sstFile := &storage.MemObject{}
+	if start.Compare(splitKey) >= 0 {
+		return nil, nil, errors.Newf("start key %s of original sst must be larger than split key %s", start, splitKey)
```
Comment (stevendanna) [inline pkg/kv/bulk/sst_batcher.go:938]: I might go with greater than rather than "larger"

**Review by stevendanna — APPROVED** [review]
Seems reasonable to me. This error really should not happen afaik so we could make these assertion failures, but I don't feel strongly about that.

**Review by dt — APPROVED** [review]

**Comment by msbutler** [comment]
TFTR!

bors r=stevendanna

**Pull request merged**